### PR TITLE
feat(AutoModeration): Support Auto Moderation

### PR DESCRIPTION
**Pull request by DonovanDMC**

### CODE REVIEW

**Review by frobinsonj — COMMENTED** [review]
Create/Modify/Delete Auto Moderation Rule endpoints support the `X-Audit-Log-Reason` header.

It should be possible to specify a reason. Not sure whether that should be put in the options or have a separate parameter.

**Comment by DonovanDMC** [comment]
> Create/Modify/Delete Auto Moderation Rule endpoints support the `X-Audit-Log-Reason` header.
> 
> It should be possible to specify a reason. Not sure whether that should be put in the options or have a separate parameter.

Done, as far as I'm aware we're actively moving to providing reasons in the `options` object if possible

File: lib/gateway/Shard.js [hunk 1]
```diff
@@ -606,6 +606,55 @@ class Shard extends EventEmitter {
 
     wsEvent(packet) {
         switch(packet.t) { /* eslint-disable no-redeclare */ // (╯°□°）╯︵ ┻━┻
```
Comment (HeadTriXz) [inline lib/gateway/Shard.js:608]: Misses the `AUTO_MODERATION_ACTION_EXECUTION` event

File: index.d.ts [hunk 2]
```diff
@@ -694,6 +743,9 @@ declare namespace Eris {
     selfVideo: boolean;
   }
   interface EventListeners {
```
Comment (HeadTriXz) [inline index.d.ts:758]: Missing `autoModerationActionExecution`

**Review by HeadTriXz — CHANGES_REQUESTED** [review]

File: index.d.ts [hunk 3]
```diff
@@ -694,6 +756,10 @@ declare namespace Eris {
...
+    autoModerationActionExecution: [guild: Guild, action: AutoModerationAction];
```
Comment (HeadTriXz) [inline index.d.ts]:
```suggestion
    autoModerationActionExecution: [guild: Guild, action: AutoModerationActionExecution];
```

**Review by HeadTriXz — CHANGES_REQUESTED** [review]

File: index.d.ts [hunk 4]
```diff
@@ -255,6 +262,61 @@ declare namespace Eris {
     permissions?: ApplicationCommandPermissions[];
   }
 
+  // Auto Moderation
+  interface AutoModerationAction {
+    metadata?: AutoModerationActionMetadata;
+    type: AutoModerationActionType;
+  }
+  interface AutoModerationActionExecution {
```
Comment (HeadTriXz) [inline index.d.ts:270]: Shouldn't this be made a class?

**Review by HeadTriXz — CHANGES_REQUESTED** [review]

File: index.d.ts (hunk 4 above)
Comment (bsian03) [reply index.d.ts:270]: Yeah probably. Or at least not make it snake_case, but a class would probably be better

Comment (DonovanDMC) [reply index.d.ts:270]: Why should it be a class? I doesn't have any specific methods and there's no other reason for this to be a class
a camel cased object would work just fine

Comment (bsian03) [reply index.d.ts:270]: Audit Log Entry is a class

Comment (HeadTriXz) [reply index.d.ts:270]: Class or not, properties should be camelCase and you could turn channel_id into a channel.

Comment (DonovanDMC) [reply index.d.ts:270]: > Audit Log Entry is a class

Existing useless classes are not a reason to make more useless classes

Comment (coolcalcacol) [reply index.d.ts:270]: Was this not the exact conversion we had with the attachment class, where we ruled that it's probably better to just have it as a class?

Comment (DonovanDMC) [reply index.d.ts:270]: Attachments are used in more than one place and are used in collections

**Pull request closed without merging**